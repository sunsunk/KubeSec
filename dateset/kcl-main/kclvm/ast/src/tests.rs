use crate::node_ref;
use crate::walker::MutSelfMutWalker;
use crate::{ast, ast::*};

/// Construct an AssignStmt node with assign_value as value
fn build_assign_node(attr_name: &str, assign_value: NodeRef<Expr>) -> NodeRef<Stmt> {
    let iden = node_ref!(Identifier {
        names: vec![Node::dummy_node(attr_name.to_string())],
        pkgpath: String::new(),
        ctx: ExprContext::Store
    });

    node_ref!(Stmt::Assign(AssignStmt {
        value: assign_value,
        targets: vec![iden],
        ty: None
    }))
}

fn get_dummy_assign_ast() -> ast::Node<ast::AssignStmt> {
    let filename = "main.k";
    let line = 1;
    let column = 1;
    let end_line = 1;
    let end_column = 2;
    ast::Node::new(
        ast::AssignStmt {
            targets: vec![Box::new(ast::Node::new(
                ast::Identifier {
                    names: vec![Node::dummy_node(String::from("a"))],
                    pkgpath: String::from(filename),
                    ctx: ast::ExprContext::Load,
                },
                String::from(filename),
                line,
                column,
                end_line,
                end_column,
            ))],
            value: Box::new(ast::Node::new(
                ast::Expr::StringLit(ast::StringLit {
                    is_long_string: false,
                    raw_value: String::from("s"),
                    value: String::from("s"),
                }),
                String::from(filename),
                line,
                column,
                end_line,
                end_column,
            )),
            ty: None,
        },
        String::from(filename),
        line,
        column,
        end_line,
        end_column,
    )
}

fn get_dummy_assign_binary_ast() -> ast::Node<ast::AssignStmt> {
    let filename = "main.k";
    let line = 1;
    let column = 1;
    let end_line = 1;
    let end_column = 2;
    ast::Node::new(
        ast::AssignStmt {
            targets: vec![Box::new(ast::Node::new(
                ast::Identifier {
                    names: vec![Node::dummy_node(String::from("a"))],
                    pkgpath: String::from(filename),
                    ctx: ast::ExprContext::Load,
                },
                String::from(filename),
                line,
                column,
                end_line,
                end_column,
            ))],
            value: Box::new(ast::Node::new(
                ast::Expr::Binary(ast::BinaryExpr {
                    op: ast::BinOp::Add,
                    left: Box::new(ast::Node::new(
                        ast::Expr::Identifier(ast::Identifier {
                            names: vec![Node::dummy_node(String::from("a"))],
                            pkgpath: String::from(filename),
                            ctx: ast::ExprContext::Load,
                        }),
                        String::from(filename),
                        line,
                        column,
                        end_line,
                        end_column,
                    )),
                    right: Box::new(ast::Node::new(
                        ast::Expr::Identifier(ast::Identifier {
                            names: vec![Node::dummy_node(String::from("a"))],
                            pkgpath: String::from(filename),
                            ctx: ast::ExprContext::Load,
                        }),
                        String::from(filename),
                        line,
                        column,
                        end_line,
                        end_column,
                    )),
                }),
                String::from(filename),
                line,
                column,
                end_line,
                end_column,
            )),
            ty: None,
        },
        String::from(filename),
        line,
        column,
        end_line,
        end_column,
    )
}

#[test]
fn test_ast_print_assign() {
    let assign_stmt = get_dummy_assign_ast();
    println!("{:?}", assign_stmt);
    let json_str = serde_json::to_string(&assign_stmt).unwrap();
    println!("{:?}", json_str);
}

#[test]
fn test_ast_print_assign_binary() {
    let assign_stmt = get_dummy_assign_binary_ast();
    println!("{:?}", assign_stmt);
    let json_str = serde_json::to_string(&assign_stmt).unwrap();
    println!("{:?}", json_str);
}

#[test]
fn test_mut_walker() {
    pub struct VarMutSelfMutWalker;
    impl<'ctx> MutSelfMutWalker<'ctx> for VarMutSelfMutWalker {
        fn walk_identifier(&mut self, identifier: &'ctx mut ast::Identifier) {
            if identifier.names[0].node == "a" {
                let id_mut = identifier.names.get_mut(0).unwrap();
                id_mut.node = "x".to_string();
            }
        }
    }
    let mut assign_stmt = get_dummy_assign_ast();
    VarMutSelfMutWalker {}.walk_assign_stmt(&mut assign_stmt.node);
    assert_eq!(assign_stmt.node.targets[0].node.names[0].node, "x")
}

#[test]
fn test_try_from_for_stringlit() {
    let str_lit = ast::StringLit::try_from("test_str".to_string()).unwrap();
    let json_str = serde_json::to_string(&str_lit).unwrap();

    let str_expected =
        r#"{"is_long_string":false,"raw_value":"\"test_str\"","value":"test_str"}"#.to_string();
    assert_eq!(str_expected, json_str);
}

#[test]
fn test_try_from_for_nameconstant() {
    let name_cons = ast::NameConstant::try_from(true).unwrap();
    let json_str = serde_json::to_string(&name_cons).unwrap();
    assert_eq!("\"True\"", json_str);

    let name_cons = ast::NameConstant::try_from(false).unwrap();
    let json_str = serde_json::to_string(&name_cons).unwrap();
    assert_eq!("\"False\"", json_str);
}

#[test]
fn test_filter_schema_with_no_schema() {
    let ast_mod = Module {
        filename: "".to_string(),
        pkg: "".to_string(),
        doc: Some(node_ref!("".to_string())),
        name: "".to_string(),
        body: vec![],
        comments: vec![],
    };
    let schema_stmts = ast_mod.filter_schema_stmt_from_module();
    assert_eq!(schema_stmts.len(), 0);
}

#[test]
fn test_filter_schema_with_one_schema() {
    let mut ast_mod = Module {
        filename: "".to_string(),
        pkg: "".to_string(),
        doc: Some(node_ref!("".to_string())),
        name: "".to_string(),
        body: vec![],
        comments: vec![],
    };
    let mut gen_schema_stmts = gen_schema_stmt(1);
    ast_mod.body.append(&mut gen_schema_stmts);
    let schema_stmts = ast_mod.filter_schema_stmt_from_module();
    assert_eq!(schema_stmts.len(), 1);
    assert_eq!(schema_stmts[0].node.name.node, "schema_stmt_0".to_string());
}

#[test]
fn test_filter_schema_with_mult_schema() {
    let mut ast_mod = Module {
        filename: "".to_string(),
        pkg: "".to_string(),
        doc: Some(node_ref!("".to_string())),
        name: "".to_string(),
        body: vec![],
        comments: vec![],
    };
    let mut gen_schema_stmts = gen_schema_stmt(10);
    ast_mod.body.append(&mut gen_schema_stmts);
    let schema_stmts = ast_mod.filter_schema_stmt_from_module();
    assert_eq!(schema_stmts.len(), 10);
    for i in 0..10 {
        assert_eq!(
            schema_stmts[i].node.name.node,
            "schema_stmt_".to_string() + &i.to_string()
        )
    }
}

#[test]
fn test_build_assign_stmt() {
    let test_expr = node_ref!(ast::Expr::Identifier(Identifier {
        names: vec![
            Node::dummy_node("name1".to_string()),
            Node::dummy_node("name2".to_string())
        ],
        pkgpath: "test".to_string(),
        ctx: ast::ExprContext::Load
    }));
    let assgin_stmt = build_assign_node("test_attr_name", test_expr);

    if let ast::Stmt::Assign(ref assign) = assgin_stmt.node {
        if let ast::Expr::Identifier(ref iden) = &assign.value.node {
            assert_eq!(iden.names.len(), 2);
            assert_eq!(iden.names[0].node, "name1".to_string());
            assert_eq!(iden.names[1].node, "name2".to_string());
            assert_eq!(iden.pkgpath, "test".to_string());
            match iden.ctx {
                ast::ExprContext::Load => {}
                _ => {
                    assert!(false);
                }
            }
        } else {
            assert!(false);
        }
    } else {
        assert!(false);
    }
}

fn gen_schema_stmt(count: i32) -> Vec<NodeRef<ast::Stmt>> {
    let mut schema_stmts = Vec::new();
    for c in 0..count {
        schema_stmts.push(node_ref!(ast::Stmt::Schema(SchemaStmt {
            doc: Some(node_ref!("".to_string())),
            name: node_ref!("schema_stmt_".to_string() + &c.to_string()),
            parent_name: None,
            for_host_name: None,
            is_mixin: false,
            is_protocol: false,
            args: None,
            mixins: vec![],
            body: vec![],
            decorators: vec![],
            checks: vec![],
            index_signature: None
        })))
    }
    schema_stmts
}
