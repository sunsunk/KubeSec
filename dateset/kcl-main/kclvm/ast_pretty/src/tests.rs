use std::path::{Path, PathBuf};

use super::print_ast_module;
use kclvm_parser::parse_file_force_errors;
use pretty_assertions::assert_eq;

const FILE_INPUT_SUFFIX: &str = ".input";
const FILE_OUTPUT_SUFFIX: &str = ".output";
const TEST_CASES: &[&str] = &[
    "arguments",
    "empty",
    "if_stmt",
    "import",
    "unary",
    "codelayout",
    "collection_if",
    "comment",
    "index_sign",
    "joined_str",
    "lambda",
    "orelse",
    "quant",
    "rule",
    "str",
    "type_alias",
    "unification",
];

fn read_data(data_name: &str) -> (String, String) {
    let mut filename = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    filename.push(
        Path::new("src")
            .join("test_data")
            .join(format!("{}{}", data_name, FILE_INPUT_SUFFIX))
            .display()
            .to_string(),
    );

    let module = parse_file_force_errors(filename.to_str().unwrap(), None);

    let mut filename_expect = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    filename_expect.push(
        Path::new("src")
            .join("test_data")
            .join(format!("{}{}", data_name, FILE_OUTPUT_SUFFIX))
            .display()
            .to_string(),
    );
    (
        print_ast_module(&module.unwrap()),
        std::fs::read_to_string(filename_expect.to_str().unwrap()).unwrap(),
    )
}

#[test]
fn test_ast_printer() {
    for case in TEST_CASES {
        let (data_input, data_output) = read_data(case);

        #[cfg(target_os = "windows")]
        let data_output = data_output.replace("\r\n", "\n");

        assert_eq!(data_input, data_output, "Test failed on {}", case);
    }
}
