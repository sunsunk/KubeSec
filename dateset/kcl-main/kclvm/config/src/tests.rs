use kclvm_version as version;
use std::{
    collections::HashMap,
    env,
    fs::{self, File},
    path::{Path, PathBuf},
};

use crate::{
    cache::{load_pkg_cache, save_pkg_cache, CacheOption},
    modfile::{get_vendor_home, KCL_PKG_PATH},
};

#[test]
fn test_vendor_home() {
    env::set_var(KCL_PKG_PATH, "test_vendor_home");
    assert_eq!(get_vendor_home(), "test_vendor_home");
    env::remove_var(KCL_PKG_PATH);

    #[cfg(target_os = "windows")]
    let root_dir = env::var("USERPROFILE").unwrap();
    #[cfg(not(target_os = "windows"))]
    let root_dir = env::var("HOME").unwrap();

    let kpm_home = PathBuf::from(root_dir)
        .join(".kcl")
        .join("kpm")
        .canonicalize()
        .unwrap();
    assert_eq!(get_vendor_home(), kpm_home.display().to_string())
}

#[test]
fn test_pkg_cache() {
    let root = PathBuf::from("./src/testdata/test_cache/")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    let mut external_pkgs = HashMap::new();
    external_pkgs.insert(
        "test_vendor".to_string(),
        "./src/testdata/test_vendor".to_string(),
    );

    let lock_path = Path::new(&root)
        .join(".kclvm/cache")
        .join(format!("{}-{}", version::VERSION, version::CHECK_SUM))
        .join("test_target");

    fs::create_dir_all(lock_path.clone()).unwrap();
    File::create(lock_path.join("test_vendor.lock")).unwrap();

    save_pkg_cache(
        &root,
        "test_target",
        "test_vendor",
        "test_data",
        CacheOption::default(),
        &external_pkgs,
    )
    .unwrap();

    assert_eq!(
        load_pkg_cache(
            &root,
            "test_target",
            "test_vendor",
            CacheOption::default(),
            &external_pkgs,
        ),
        Some("test_data".to_string())
    )
}
