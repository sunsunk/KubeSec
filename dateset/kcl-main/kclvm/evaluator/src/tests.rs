use crate::Evaluator;
use kclvm_loader::{load_packages, LoadPackageOptions};
use kclvm_parser::LoadProgramOptions;

#[macro_export]
macro_rules! evaluator_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            let p = load_packages(&LoadPackageOptions {
                paths: vec!["test.k".to_string()],
                load_opts: Some(LoadProgramOptions {
                    k_code_list: vec![$src.to_string()],
                    ..Default::default()
                }),
                load_builtin: false,
                ..Default::default()
            })
            .unwrap();
            let evaluator = Evaluator::new(&p.program);
            insta::assert_snapshot!(format!("{}", evaluator.run().unwrap().1));
        }
    };
}

evaluator_snapshot! {expr_stmt_0, "1"}
evaluator_snapshot! {expr_stmt_1, "2.0"}
evaluator_snapshot! {expr_stmt_2, "True"}
evaluator_snapshot! {expr_stmt_3, r#"None"#}
evaluator_snapshot! {expr_stmt_4, r#"[1, 2, 3]"#}
evaluator_snapshot! {expr_stmt_5, r#"{k = "v"}"#}

evaluator_snapshot! {assign_stmt_0, "a = 1"}
evaluator_snapshot! {assign_stmt_1, "a = 1 + 1"}
evaluator_snapshot! {assign_stmt_2, "a = (1 + 2)"}
evaluator_snapshot! {assign_stmt_3, r#"a = 1
b = a + 1
"#}
evaluator_snapshot! {assign_stmt_4, r#"a: int = 1
b: int = a + 1
"#}

evaluator_snapshot! {aug_assign_stmt_0, r#"_a = 1
_a += 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_1, r#"_a = 1
_a -= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_2, r#"_a = 1
_a *= 2
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_3, r#"_a = 2
_a /= 2
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_4, r#"_a = 3
_a %= 2
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_5, r#"_a = 3
_a **= 2
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_6, r#"_a = 3
_a <<= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_7, r#"_a = 3
_a >>= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_8, r#"_a = 3
_a |= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_9, r#"_a = 3
_a ^= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_10, r#"_a = 3
_a &= 1
a = _a
"#}
evaluator_snapshot! {aug_assign_stmt_11, r#"_a = 3
_a //= 2
a = _a
"#}

evaluator_snapshot! {assert_stmt_0, r#"assert True, "msg"
a = 1
"#}

evaluator_snapshot! {assert_stmt_1, r#"assert False if False, "msg"
a = 1
"#}

evaluator_snapshot! {if_stmt_0, r#"if True:
    a = 1
else:
    b = 2
"#}
evaluator_snapshot! {if_stmt_1, r#"if False:
    a = 1
else:
    b = 2
"#}
evaluator_snapshot! {if_stmt_3, r#"if False:
    a = 1
elif True:
    b = 2
else:
    c = 3
"#}
evaluator_snapshot! {if_stmt_4, r#"if False:
    a = 1
elif False:
    b = 2
else:
    c = 3
"#}
evaluator_snapshot! {if_stmt_5, r#"if False:
    a = 1
else:
    if True:
        b = 2
    else:
        c = 3
"#}

evaluator_snapshot! {import_stmt_0, r#"import math
a = 1
"#}
evaluator_snapshot! {import_stmt_1, r#"import math
import math
b = 2
"#}

evaluator_snapshot! {quant_expr_0, r#"b = all a in [1, 2, 3] {
    a > 0
}
"#}
evaluator_snapshot! {quant_expr_1, r#"b = any a in [1, 2, 3] {
    a > 2
}
"#}
evaluator_snapshot! {quant_expr_2, r#"b = all a in [1, 2, 3] {
    a > 5
}
"#}
evaluator_snapshot! {quant_expr_3, r#"b = any a in [1, 2, 3] {
    a > 5
}
"#}
evaluator_snapshot! {quant_expr_4, r#"b = map a in [1, 2, 3] {
    a + 1
}
"#}
evaluator_snapshot! {quant_expr_5, r#"b = filter a in [1, 2, 3] {
    a > 1
}
"#}
evaluator_snapshot! {quant_expr_6, r#"b = map a in [1, 2, 3] {
    a ** 2
}
"#}
evaluator_snapshot! {quant_expr_7, r#"b = filter a in [1, 2, 3] {
    a == 0
}
"#}

evaluator_snapshot! {if_expr_0, r#"a = 1 if True else 0"#}
evaluator_snapshot! {if_expr_1, r#"a = 1 if False else 0"#}
evaluator_snapshot! {if_expr_2, r#"a = 1 if False else 0 if False else 2"#}

evaluator_snapshot! {unary_expr_0, r#"a = +1"#}
evaluator_snapshot! {unary_expr_1, r#"a = -1"#}
evaluator_snapshot! {unary_expr_2, r#"a = ~1"#}
evaluator_snapshot! {unary_expr_3, r#"a = not None"#}

evaluator_snapshot! {binary_expr_0, r#"a = 1 + 1 * 2 - 4"#}
evaluator_snapshot! {binary_expr_1, r#"a = None or {}
b = [] and {}
"#}

evaluator_snapshot! {selector_expr_0, r#"a = {k = "v"}.k
b = {k = "v"}?.k
c = None?.k
"#}
evaluator_snapshot! {selector_expr_1, r#"a = [1, 2, 3]?[::-1]
b = a?[-1]
c = a?[0]
d = None?[1]
"#}

evaluator_snapshot! {subscript_expr_0, r#"a = [1, 2, 3][::-1]
b = a[-1]
c = a[0]
"#}
evaluator_snapshot! {subscript_expr_1, r#"a = [1, 2, 3]?[::-1]
b = a?[-1]
c = a?[0]
d = None?[1]
"#}

evaluator_snapshot! {compare_expr_0, r#"a = 1 < 10
b = 1 < 10 < 100
c = 1 > 10 > 100
d = 1 < 10 > 100
"#}

evaluator_snapshot! {paren_expr_0, r#"a = 2 * (1 + 1)
b = (((1 + 1))) * 2
"#}

evaluator_snapshot! {list_expr_0, r#"a = [1, 2, 3]
b = [1, if True: 2, 3]
c = [1, if False: 2, 3]
d = [1, *[2, 3]]
"#}

evaluator_snapshot! {dict_expr_0, r#"a = {k1 = "v1", k2 = "v2"}
b = {k1 = "v1", if True: k2 = "v2"}
c = {k1 = "v1", if False: k2 = "v2"}
d = {k1 = "v1", **{k2 = "v2"}}
"#}

evaluator_snapshot! {loop_0, r#"a = [i ** 2 for i in [1, 2, 3]]"#}
evaluator_snapshot! {loop_1, r#"a = [i + j for i in [1, 2, 3] for j in [1, 2, 3] if i < j]"#}

evaluator_snapshot! {literal_0, r#"longStringStartWithNewline = """\
This is the second line
This is the third line
"""
"#}
evaluator_snapshot! {literal_1, r#"a = {k = "v"}
b = "${a: #json}"
"#}
evaluator_snapshot! {literal_2, r#"a = 1Mi
b = 2K
"#}

evaluator_snapshot! {lambda_0, r#"f = lambda x {x * 2}
a = f(1)
b = f(2)
"#}
evaluator_snapshot! {lambda_1, r#"a = lambda x {x * 2}(1)
b = lambda x {x * 2}(2)
"#}
evaluator_snapshot! {lambda_2, r#"import math
a = math.log(10)
b = len("abc")
c = len([1, 2])
"#}
evaluator_snapshot! {lambda_3, r#"
x = lambda {
    a = 1
    lambda {
        a + 1
    }()
}()
"#}
evaluator_snapshot! {lambda_4, r#"
x = lambda {
    a = 1
    b = 2
    lambda x {
        a + b + x
    }(3)
}()
"#}
evaluator_snapshot! {lambda_5, r#"
func = lambda config: {str:} {
    x = 1
    lambda {
        y = 1
        lambda {
            z = 1
            lambda {
                {value = x + y + z + config.key}
            }()
        }()
    }()
}

x = func({key = 1})
"#}

evaluator_snapshot! {schema_0, r#"
schema Person:
    name: str = "Alice"
    age: int = 10

alice = Person {}
bob = Person {
    name = "Bob"
    age = 18
}
"#}
