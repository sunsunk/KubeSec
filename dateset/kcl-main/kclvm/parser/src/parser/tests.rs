use crate::parse_file_force_errors;
use regex::Regex;

#[test]
fn test_parse_file_not_found() {
    match parse_file_force_errors("The file path is invalid", None) {
        Ok(_) => {
            panic!("unreachable")
        }
        Err(err_msg) => {
            assert!(
                Regex::new(r"^Failed to load KCL file 'The file path is invalid'. Because.*")
                    .unwrap()
                    .is_match(&err_msg.to_string())
            );
        }
    }
}
