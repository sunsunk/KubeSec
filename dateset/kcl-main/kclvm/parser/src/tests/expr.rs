use crate::tests::parse_expr_snapshot;

parse_expr_snapshot!(smoke_test_parsing_expr_0, "1\n");
parse_expr_snapshot!(smoke_test_parsing_expr_1, "\"1\"\n");
parse_expr_snapshot!(named_literal_expr_0, "Undefined");
parse_expr_snapshot!(named_literal_expr_1, "None");
parse_expr_snapshot!(named_literal_expr_2, "True");
parse_expr_snapshot!(named_literal_expr_3, "False");
parse_expr_snapshot!(nonstring_literal_expr, r####"1234"####);
parse_expr_snapshot!(string_literal_expr_0, r####"'1234'"####);
parse_expr_snapshot!(string_literal_expr_1, r####""1234""####);
parse_expr_snapshot!(string_literal_expr_2, r####""1234\n""####);
parse_expr_snapshot!(number_bin_suffix_expr, r####"1234Ki"####);
parse_expr_snapshot!(unary_expr, r####"+1"####);
parse_expr_snapshot!(binary_expr_0, r####"1+2+3"####);
parse_expr_snapshot!(binary_expr_1, r####"1+2*3-4"####);
parse_expr_snapshot!(binary_expr_2, r####"1+2*3/4"####);
parse_expr_snapshot!(binary_expr_3, r####"a or b"####);
parse_expr_snapshot!(binary_expr_4, r####"x == a or b"####);
parse_expr_snapshot!(binary_expr_5, r####"22 > 11 and 111 < 222"####);
parse_expr_snapshot!(binary_expr_6, r####"int(e.value) > 1 and i == 0"####);
parse_expr_snapshot!(binary_expr_7, r####"key in ['key']"####);
parse_expr_snapshot!(binary_expr_8, r####"key not in ['key']"####);
parse_expr_snapshot!(binary_expr_9, r####"1 is 1 and 11 is not 22"####);
parse_expr_snapshot!(binary_expr_10, r####"1 + a and b"####);
parse_expr_snapshot!(binary_expr_with_paren, r####"1*(2+3)-4"####);
parse_expr_snapshot!(logic_expr_0, r####"0 < a < 100"####);
parse_expr_snapshot!(logic_expr_1, r####"0 < a < 100 + a"####);
parse_expr_snapshot!(logic_expr_2, r####"100 > a > 0"####);
parse_expr_snapshot!(logic_expr_3, r####"100 + a > a > 0"####);
parse_expr_snapshot!(logic_expr_4, r####"a is b"####);
parse_expr_snapshot!(logic_expr_5, r####"a is not True"####);
parse_expr_snapshot!(logic_expr_6, r####"not False or a > 0 and b is True"####);
parse_expr_snapshot!(if_expr, r####"1 if true else 2"####);
parse_expr_snapshot!(primary_expr_0, r####"a.b.c"####);
parse_expr_snapshot!(primary_expr_1, r####"'{}'.format(1)"####);
parse_expr_snapshot!(primary_expr_2, r####"str(1).isdigit()"####);
parse_expr_snapshot!(list_expr_0, r####"[1, 2, 3]"####);
parse_expr_snapshot!(list_expr_1, r####"[1, if True: 2, 3]"####);
parse_expr_snapshot!(list_comp_expr_0, r####"[x ** 2 for x in [1, 2, 3]]"####);
parse_expr_snapshot!(list_comp_expr_1, r####"[i for i in [1, 2, 3] if i > 2]"####);
parse_expr_snapshot!(dict_expr, r####"{k0=v0, k1=v1}"####);
parse_expr_snapshot!(
    dict_comp_expr,
    r####"{k: v + 1 for k, v in {k1 = 1, k2 = 2}}"####
);
parse_expr_snapshot!(subscript_expr_0, r####"a[0]"####);
parse_expr_snapshot!(subscript_expr_1, r####"b["k"]"####);
parse_expr_snapshot!(subscript_expr_2, r####"c?[1]"####);
parse_expr_snapshot!(subscript_expr_3, r####"a[1:]"####);
parse_expr_snapshot!(subscript_expr_4, r####"a[:-1]"####);
parse_expr_snapshot!(subscript_expr_5, r####"a[1:len]"####);
parse_expr_snapshot!(subscript_expr_6, r####"a[0:-1]"####);
parse_expr_snapshot!(subscript_expr_7, r####"a[::]"####);
parse_expr_snapshot!(subscript_expr_8, r####"a[1::]"####);
parse_expr_snapshot!(subscript_expr_9, r####"a[:0:]"####);
parse_expr_snapshot!(subscript_expr_10, r####"a[::-1]"####);
parse_expr_snapshot!(subscript_expr_11, r####"a[1::2]"####);
parse_expr_snapshot!(subscript_expr_12, r####"a[:2:1]"####);
parse_expr_snapshot!(subscript_expr_13, r####"a[1:2:]"####);
parse_expr_snapshot!(subscript_expr_14, r####"a[1:3:1]"####);
parse_expr_snapshot!(call_expr_0, r####"func0()"####);
parse_expr_snapshot!(call_expr_1, r####"func1(1)"####);
parse_expr_snapshot!(call_expr_2, r####"func2(x=2)"####);
parse_expr_snapshot!(call_expr_3, r####"func3(1,x=2)"####);
parse_expr_snapshot!(quant_expr_0, r####"all x in collection {x > 0}"####);
parse_expr_snapshot!(quant_expr_1, r####"any y in collection {y < 0}"####);
parse_expr_snapshot!(quant_expr_2, r####"map x in collection {x + 1}"####);
parse_expr_snapshot!(quant_expr_3, r####"filter x in collection {x > 1}"####);
parse_expr_snapshot!(quant_expr_4, r####"filter x in collection {x > 1}"####);
parse_expr_snapshot!(
    quant_expr_5,
    r####"map i, e in [{k1 = "v1", k2 = "v2"}] { e }"####
);
parse_expr_snapshot!(
    quant_expr_6,
    r####"map i, e in [{k1 = "v1", k2 = "v2"}] { e if i > 0 }"####
);
parse_expr_snapshot!(lambda_expr_0, r####"lambda {}"####);
parse_expr_snapshot!(lambda_expr_1, r####"lambda x {}"####);
parse_expr_snapshot!(lambda_expr_2, r####"lambda x: int -> int {x}"####);
parse_expr_snapshot!(
    lambda_expr_3,
    r####"lambda {
    if True:
        _a = 1
    else:
        _a = 2
    _a
}"####
);
parse_expr_snapshot!(
    config_expr_0,
    r####"{
    "name" = {
        "name": "alice"
    },
    "gender" = "female"
}"####
);
parse_expr_snapshot!(
    config_expr_1,
    r####"{
    "name" = {
        "name": "alice"
    }
    "gender" = "female",
}"####
);
parse_expr_snapshot!(
    config_expr_2,
    r####"{
    "name" = {
        "name": "alice",
    }
    "gender" = "female"
}"####
);
parse_expr_snapshot!(
    config_if_expr_0,
    r####"{
    if True:
        a = 1
}"####
);
parse_expr_snapshot!(
    config_if_expr_1,
    r####"{
    if True:
        a = 1
    else:
        a = 2
}"####
);
parse_expr_snapshot!(
    config_if_expr_2,
    r####"{
    if True:
        a = 1
    elif x > 1:
        a = 2
    else:
        a = 3
}"####
);
parse_expr_snapshot!(
    config_if_expr_3,
    r####"{
    if True:
        if False:
            a = 1
}"####
);
parse_expr_snapshot!(schema_expr_0, r####"Schema {}"####);
parse_expr_snapshot!(schema_expr_1, r####"Schema {k=v}"####);
parse_expr_snapshot!(schema_expr_2, r####"Schema () {k=v}"####);
parse_expr_snapshot!(schema_expr_3, r####"Schema (1, 2) {k=v}"####);
parse_expr_snapshot!(
    schema_expr_4,
    r####"Schema (1, 2) {
    k=v
}"####
);
parse_expr_snapshot!(
    line_continue,
    r####"1 + \
2
"####
);
parse_expr_snapshot!(parse_joined_string_0, r####"'${123+200}'"####);
parse_expr_snapshot!(parse_joined_string_1, r####"'abc${a+1}cde'"####);
parse_expr_snapshot!(expr_with_paren_0, r####"(2+3)"####);
parse_expr_snapshot!(expr_with_paren_1, r####"((2+3)"####);
parse_expr_snapshot!(expr_with_paren_2, r####"(2+3))"####);
parse_expr_snapshot!(expr_with_bracket_0, r####"[2,3]"####);
parse_expr_snapshot!(expr_with_bracket_1, r####"[[2,3]"####);
parse_expr_snapshot!(expr_with_bracket_2, r####"[2,3]]"####);
parse_expr_snapshot!(expr_with_bracket_3, r####"[2,3"####);
parse_expr_snapshot!(expr_with_bracket_4, r####"["####);
parse_expr_snapshot!(
    expr_with_bracket_5,
    r####"[
    1
    2,
]
        "####
);
parse_expr_snapshot!(
    expr_with_bracket_6,
    r####"[
    1,2,
]
        "####
);
parse_expr_snapshot!(
    expr_with_bracket_7,
    r####"[
    1,2,

        "####
);
parse_expr_snapshot!(expr_with_brace_0, r####"{a=2}"####);
parse_expr_snapshot!(expr_with_brace_1, r####"{a=2}}"####);
parse_expr_snapshot!(expr_with_delim_0, r####"({a=2}"####);
parse_expr_snapshot!(expr_with_delim_1, r####"({a=(2}"####);
parse_expr_snapshot!(expr_with_delim_2, r####"{a=[2]"####);
parse_expr_snapshot!(expr_with_delim_3, r####"[{a=2}"####);
parse_expr_snapshot!(expr_with_delim_4, r####"({a=[2}"####);
parse_expr_snapshot!(expr_with_delim_5, r####"{"####);
parse_expr_snapshot!(
    expr_with_delim_6,
    r####"{
    a = 1
}"####
);
parse_expr_snapshot!(
    expr_with_delim_7,
    r####"{
    a = 1
"####
);
