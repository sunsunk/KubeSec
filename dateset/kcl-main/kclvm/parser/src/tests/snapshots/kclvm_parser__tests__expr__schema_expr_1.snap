---
source: parser/src/tests/expr.rs
expression: "crate::tests::parsing_expr_string(r####\"Schema {k=v}\"####)"
---
Node {
    node: Schema(
        SchemaExpr {
            name: Node {
                node: Identifier {
                    names: [
                        Node {
                            node: "Schema",
                            filename: "",
                            line: 1,
                            column: 0,
                            end_line: 1,
                            end_column: 6,
                        },
                    ],
                    pkgpath: "",
                    ctx: Load,
                },
                filename: "",
                line: 1,
                column: 0,
                end_line: 1,
                end_column: 6,
            },
            args: [],
            kwargs: [],
            config: Node {
                node: Config(
                    ConfigExpr {
                        items: [
                            Node {
                                node: ConfigEntry {
                                    key: Some(
                                        Node {
                                            node: Identifier(
                                                Identifier {
                                                    names: [
                                                        Node {
                                                            node: "k",
                                                            filename: "",
                                                            line: 1,
                                                            column: 8,
                                                            end_line: 1,
                                                            end_column: 9,
                                                        },
                                                    ],
                                                    pkgpath: "",
                                                    ctx: Load,
                                                },
                                            ),
                                            filename: "",
                                            line: 1,
                                            column: 8,
                                            end_line: 1,
                                            end_column: 9,
                                        },
                                    ),
                                    value: Node {
                                        node: Identifier(
                                            Identifier {
                                                names: [
                                                    Node {
                                                        node: "v",
                                                        filename: "",
                                                        line: 1,
                                                        column: 10,
                                                        end_line: 1,
                                                        end_column: 11,
                                                    },
                                                ],
                                                pkgpath: "",
                                                ctx: Load,
                                            },
                                        ),
                                        filename: "",
                                        line: 1,
                                        column: 10,
                                        end_line: 1,
                                        end_column: 11,
                                    },
                                    operation: Override,
                                    insert_index: -1,
                                },
                                filename: "",
                                line: 1,
                                column: 8,
                                end_line: 1,
                                end_column: 11,
                            },
                        ],
                    },
                ),
                filename: "",
                line: 1,
                column: 7,
                end_line: 1,
                end_column: 12,
            },
        },
    ),
    filename: "",
    line: 1,
    column: 0,
    end_line: 1,
    end_column: 12,
}

